Data Structures/Linked-Lists/Implementations of single linked list.cpp

```diff
@@ -0,0 +1,346 @@
+}
+/*End of search()*/
+
+struct node *addAtBeg(struct node *start,int data)
+{
+	struct node *tmp;
+	tmp=(struct node *)malloc(sizeof(struct node));
+	tmp->info=data;
+	tmp->link=start;
+	start=tmp;
+	return start;
+}
+/*End of addAtBeg()*/
+
+struct node *addAtEnd(struct node *start,int data)
+{
+	struct node *p,*tmp;
+	tmp=(struct node *)malloc(sizeof(struct node));
+	tmp->info=data;
+	p=start;
+	while(p->link!=NULL)
+		p=p->link;
+	p->link=tmp;
+	tmp->link=NULL;
+	return start;
+}
+/*End of addAtEnd()*/
+
+struct node *addAfter(struct node *start,int data,int item)
+{
+	struct node *tmp,*p;
+	p=start;
+	while(p!=NULL)
+	{
+		if(p->info==item)
+		{
+			tmp=(struct node *)malloc(sizeof(struct node));
+			tmp->info=data;
+			tmp->link=p->link;
+			p->link=tmp;
+			return start;
+		}
+		p=p->link;
+	}
+	printf("%d not present in the list\n",item);
+	return start;
+}
+/*End of addafter()*/
+
+struct node *addBefore(struct node *start,int data,int item)
+{
+	struct node *tmp,*p;
+	if(start==NULL )
+	{
+		printf("List is empty\n");
+		return start;
+	}	
+	/*If data to be inserted before first node*/
+	if(item==start->info)
+	{
+		tmp=(struct node *)malloc(sizeof(struct node));
+		tmp->info=data;
+		tmp->link=start;
+		start=tmp;
+		return start;
+	}	
+	p=start;
+	while(p->link!=NULL) 
+	{
+		if(p->link->info==item)
+		{
+			tmp=(struct node *)malloc(sizeof(struct node));
+			tmp->info=data;
+			tmp->link=p->link;
+			p->link=tmp;
+			return start;
+		}
+		p=p->link;
+	}
+	printf("%d not present in the list\n",item);
+	return start;
+}
+/*End of addBefore()*/	
+
+struct node *addAtPos(struct node *start,int data,int pos)
+{
+	struct node *tmp,*p;
+	int i;
+	tmp=(struct node *)malloc(sizeof(struct node));
+	tmp->info=data;
+	if(pos==1)
+	{
+		tmp->link=start; 
+		start=tmp;
+		return start;
+	}
+	p=start;
+	for(i=1; i<pos-1 && p!=NULL; i++)
+		p=p->link;
+	if(p==NULL)
+		printf("There are less than %d elements\n",pos);
+	else
+	{
+		tmp->link=p->link;
+		p->link=tmp;
+	}	
+	return start;
+}
+/*End of addAtPos()*/
+
+struct node *del(struct node *start,int data)
+{
+	struct node *tmp,*p;
+	if(start==NULL)
+	{
+		printf("List is empty\n");
+		return start;
+	}
+	/*Deletion of first node*/
+	if(start->info==data)
+	{
+		tmp=start;
+		start=start->link;  
+		free(tmp);
+		return start;
+	}
+	/*Deletion in between or at the end*/
+	p=start;
+	while(p->link!=NULL)
+	{
+		if(p->link->info==data)   
+		{
+			tmp=p->link;
+			p->link=tmp->link;
+			free(tmp);
+			return start;
+		}
+		p=p->link;
+	}
+	printf("Element %d not found\n",data);
+	return start;
+}
+/*End of del()*/
+
+struct node *reverse(struct node *start)
+{
+	struct node *prev, *ptr, *next;
+	prev=NULL;
+   	ptr=start;
+	while(ptr!=NULL)
+	{
+		next=ptr->link;
+		ptr->link=prev;
+		prev=ptr;
+		ptr=next;
+	}
+	start=prev;
+	return start;
+}
+/*End of reverse()*/
+
+/*  In whole program there is only one loop many times in different block os time complexity will be k*O(n)=O(n)  */ 

```

Comment:
Please explain the **time complexity for each of the functions** separately as comments above the function. The time complexity of the entire program is irrelevant.

Reply:
@rishabhgarg25699 Please take this also into consideration.